src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,752 @@
+  // each other, we combine their equivalence classes and choose a new class
+  // representative to use to disambiguate the groups.
+  DisjointSets equivalenceClasses;
+
+  // When we see a new group, we have to make sure its shape doesn't conflict
+  // with any existing group. If it does conflict, we need to update its shape
+  // and try again. Maintain a stack of group indices whose shapes we need to
+  // check for uniqueness to avoid deep recursions.
+  std::vector<size_t> shapesToUpdate;
+
+  void run(Module* module) override {
+    // There are no recursion groups to minimize if GC is not enabled.
+    if (!module->features.hasGC()) {
+      return;
+    }
+
+    initBrandOptions();
+
+    types = ModuleUtils::getPrivateHeapTypes(*module);
+    for (auto type : ModuleUtils::collectHeapTypes(*module)) {
+      typeIndices.insert({type, typeIndices.size()});
+    }
+
+    // The number of types to optimize is an upper bound on the number of
+    // recursion groups we will emit.
+    groups.reserve(types.size());
+
+    // Compute the strongly connected components and ensure they form distinct
+    // recursion groups.
+    for (auto scc : TypeSCCs(types)) {
+      [[maybe_unused]] size_t index = equivalenceClasses.addSet();
+      assert(index == groups.size());
+      groups.emplace_back();
+
+      // The SCC is not necessarily topologically sorted to have the supertypes
+      // come first. Fix that.
+      std::vector<HeapType> sccTypes(scc.begin(), scc.end());
+      auto deps = createSubtypeGraph(sccTypes);
+      auto permutation = *TopologicalOrders(deps).begin();
+      groups.back().group.resize(sccTypes.size());
+      for (size_t i = 0; i < sccTypes.size(); ++i) {
+        groups.back().group[i] = sccTypes[permutation[i]];
+      }
+      assert(shapesToUpdate.empty());
+      shapesToUpdate.push_back(index);
+      updateShapes();
+    }
+
+    rewriteTypes(*module);
+  }
+
+  void initBrandOptions() {
+    // Initialize the field options for brand types lazily here to avoid
+    // depending on global constructor ordering.
+    [[maybe_unused]] static bool fieldsInitialized = []() {
+      BrandTypeIterator::initFieldOptions();
+      return true;
+    }();
+  }
+
+  void updateShapes() {
+    while (!shapesToUpdate.empty()) {
+      auto index = shapesToUpdate.back();
+      shapesToUpdate.pop_back();
+      updateShape(index);
+    }
+  }
+
+  void updateShape(size_t group) {
+    auto [it, inserted] =
+      groupShapeIndices.insert({RecGroupShape(groups[group].group), group});
+    if (inserted) {
+      // This shape was unique. We're done.
+      return;
+    }
+
+    // We have a conflict. There are five possibilities:
+    //
+    //   1. We are trying to insert the next permutation of an existing
+    //      equivalence class and have found that...
+    //
+    //     A. The next permutation is an automorphism of some previous
+    //        permutation of the same equivalence class.
```

Comment:
IIUC when we fail to insert into `groupShapeIndices` that means that we have seen equality of `RecGroupShape`s. IIUC `RecGroupShape` compares the shape of the types it is given, ignoring type identity and just using the shape of the types + their relations to each other, which are done using the internal indices; hence, the order matters.

If so, then if we fail to insert that means that we have the exact shape of another `RecGroupShape`, including order.

What does "We are trying to insert the next permutation of an existing equivalence class" mean, in that context? We are not looking at equivalence classes that ignore the order. And why does this sentence mention "next permutation"?

I would expect this code to say something like "Oh well, looks like that exact rec group shape was seen before. Let's modify ourselves somehow to avoid that. We can do that by either permuting ourselves, or by modifying the brand. Let's pick which of those two to try first based on some heuristic, as follows [..]".

What am I missing?

Reply:
To summarize an offline conversation, what I was missing was that the code here also tries to memoize all seen permutations of rec groups. We don't just see a conflict and fix it and move on, but note all such conflicts so that later potential conflicts with the same isomorphic shape do not end up traversing the same path as we do, and repeat the work. (My sense is that some kind of pseudo random permutation/brand adjustment would also avoid lengthy repetition but I don't have a guess if that would be simpler.)